File: ADSMOD/client/src/services/browser.ts
import { API_BASE_URL } from '../constants';
import { fetchWithTimeout, extractErrorMessage, HTTP_TIMEOUT } from './http';

export async function fetchTableList(): Promise<{ tables: { table_name: string; display_name: string; category: string }[]; error: string | null }> {
    try {
        const response = await fetchWithTimeout(
            `${API_BASE_URL}/browser/tables`,
            { method: 'GET' },
            HTTP_TIMEOUT
        );

        if (!response.ok) {
            const data = await response.json().catch(() => ({}));
            const message = extractErrorMessage(response, data);
            return { tables: [], error: message };
        }

        const data = await response.json();
        return { tables: data.tables || [], error: null };
    } catch (error) {
        if (error instanceof Error) {
            return { tables: [], error: error.message };
        }
        return { tables: [], error: 'An unknown error occurred.' };
    }
}

export async function fetchTableData(
    tableName: string,
    limit = 50,
    offset = 0
): Promise<{
    data: Record<string, unknown>[];
    columns: string[];
    rowCount: number;
    totalRows: number;
    columnCount: number;
    displayName: string;
    error: string | null;
}> {
    try {
        const response = await fetchWithTimeout(
            `${API_BASE_URL}/browser/data/${encodeURIComponent(tableName)}?limit=${limit}&offset=${offset}`,
            { method: 'GET' },
            HTTP_TIMEOUT
        );

        if (!response.ok) {
            const data = await response.json().catch(() => ({}));
            const message = extractErrorMessage(response, data);
            return {
                data: [],
                columns: [],
                rowCount: 0,
                totalRows: 0,
                columnCount: 0,
                displayName: '',
                error: message,
            };
        }

        const result = await response.json();
        return {
            data: result.data || [],
            columns: result.columns || [],
            rowCount: result.row_count || 0,
            totalRows: result.total_rows || 0,
            columnCount: result.column_count || 0,
            displayName: result.display_name || tableName,
            error: null,
        };
    } catch (error) {
        if (error instanceof Error) {
            return {
                data: [],
                columns: [],
                rowCount: 0,
                totalRows: 0,
                columnCount: 0,
                displayName: '',
                error: error.message,
            };
        }
        return {
            data: [],
            columns: [],
            rowCount: 0,
            totalRows: 0,
            columnCount: 0,
            displayName: '',
            error: 'An unknown error occurred.',
        };
    }
}

File: ADSMOD/server/routes/browser.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, status

from ADSMOD.server.entities.browser import (
    TableDataResponse,
    TableInfo,
    TableListResponse,
)
from ADSMOD.server.repositories.database import database
from ADSMOD.server.repositories.serialization.data import DataSerializer
from ADSMOD.server.common.constants import (
    BROWSER_DATA_ENDPOINT,
    BROWSER_ROUTER_PREFIX,
    BROWSER_TABLE_CATEGORIES,
    BROWSER_TABLE_DISPLAY_NAMES,
    BROWSER_TABLES_ENDPOINT,
)

from ADSMOD.server.common.utils.logger import logger

serializer = DataSerializer()
router = APIRouter(prefix=BROWSER_ROUTER_PREFIX, tags=["browser"])


###############################################################################
@router.get(
    BROWSER_TABLES_ENDPOINT,
    response_model=TableListResponse,
    status_code=status.HTTP_200_OK,
)
async def list_tables() -> TableListResponse:
    tables = [
        TableInfo(
            table_name=table_name,
            display_name=display_name,
            category=BROWSER_TABLE_CATEGORIES[table_name],
        )
        for table_name, display_name in BROWSER_TABLE_DISPLAY_NAMES.items()
    ]
    return TableListResponse(tables=tables)


###############################################################################
@router.get(
    f"{BROWSER_DATA_ENDPOINT}/{{table_name}}",
    response_model=TableDataResponse,
    status_code=status.HTTP_200_OK,
)
async def get_table_data(
    table_name: str,
    limit: int = 50,
    offset: int = 0,
) -> TableDataResponse:
    if table_name not in BROWSER_TABLE_DISPLAY_NAMES:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Table '{table_name}' not found or not available for browsing.",
        )

    try:
        total_rows = database.count_rows(table_name)
        df = serializer.load_table(table_name, limit=limit, offset=offset)
    except Exception as exc:  # noqa: BLE001
        logger.exception("Failed to load table %s", table_name)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to load table data: {exc}",
        ) from exc

    # Convert DataFrame to list of dicts for JSON response
    columns = df.columns.tolist()
    data = df.fillna("").to_dict(orient="records")

    return TableDataResponse(
        table_name=table_name,
        display_name=BROWSER_TABLE_DISPLAY_NAMES[table_name],
        total_rows=total_rows,
        row_count=len(df),
        column_count=len(columns),
        columns=columns,
        data=data,
    )
